### DISCUSSION TITLE
Sexp (html style) for creating clog nodes declaratively.

### DISCUSSION THREAD

**Discussion by Inc0n** [rabbibotton/clog#241] [Show and tell]
I have made a sexp mark up language for creating nested clog nodes easier. Tell me yours thoughts!
In your `on-new-window` function, you can now do:
```
(sexp->clog-nodes
 body
 `(div ()
	   (h1 () "Example App")
	   (h4 () "Some sub header")
	   ,@(loop for str in '("a" "b" "c")
			   collect `(p () ,str))
	   (h4 () "Next sub header")
	   ,@(loop for str in '("d" "e" "f")
			   collect `(p () ,str))))
```
As suppose to the nested `let` `create-<tag>`s.
Basically, the syntax follows that `(<tag> (<attri-name> <attri-value> ...) <rest-nodes-or-string> ...)`.

<details>
<summary>Code</summary>

```
(defparameter *html-tag-clog-type-alist*
  '(((div) . clog-div)
	((address article aside header main nav
	  p pre section blockquote h1 h2 h3 h4 h5 h6
	  hgroup)
	 . clog-section)
	((ul) . clog-unordered-list)
	((li) . clog-list-item)
	((p) . clog-p)
	((a) . clog-a)))

(defun create-node (parent tag content attrs
					&key (style nil)
					  (hidden nil)
					  (class nil)
					  (html-id nil)
					  (auto-place t))
  (create-child parent
				(format nil "<~A~@[~A~]~@[~A~]>~A</~A>"
						tag
						(format nil "~{ ~a~}"
								(loop for (name val) on attrs by #'cddr
									  collect (format nil "~A='~A'" name val)))
						;; (when class
                        ;;   (format nil " class='~A'"
                        ;;           (escape-string class)))
						(when (or hidden style)
						  (format nil " style='~@[~a~]~@[~a~]'"
								  (when hidden "visibility:hidden;")
								  style))
						(escape-string content)
						tag)
				:clog-type (or (cdr (assoc tag *html-tag-clog-type-alist*
										   :test #'member))
							   (error "Tag ~a is not implemented for clog-type" tag))
				:html-id    html-id
				:auto-place auto-place))

(defun take/drop (pred lst)
  (loop for x in lst
		if (funcall pred x)
		  collect x into takes
		else collect x into drops
		finally (return (values takes drops))))

(defun sexp->clog-nodes (root sexp)
  (destructuring-bind (tag attrs &rest body)
	  sexp
	(multiple-value-bind (contents children)
		(take/drop #'stringp body)
	  (let ((node (create-node
				   root tag
				   (format nil "~{~a~}" contents)
				   attrs)))
		(dolist (child children)
		  (sexp->clog-nodes node child))))))
```

</details>

Note: Because it uses `*html-tag-clog-type-alist*` to translate the tag into clog-type, therefore, I hadn't coded all other tag->type exhaustively (only the ones I need so far).

Also because attributes are directly translated into strings, so for example, create-a accepts :link, but here you would use :href instead.

**Comment by rabbibotton** [comment]
Cool. Have you seen Tutorial 33 - https://github.com/rabbibotton/clog/blob/main/tutorial/33-tutorial.lisp and with-clog-create?

What are the key advantages over with-clog-create?

**Reply by Inc0n** [reply]
Ahh, I should have known. So `with-clog-create` is definitely talored to clog, it flows very neetly. In comparison, I made mine with a different approach in mind. 
1. :content is made as strings in the children nodes list,
2. the attribute list is made into as part of the html string, 
3. it invokes `create-node` as suppose to `create-<tag>`.
